pkg/util/errors.go

```diff
@@ -186,3 +192,145 @@ func HandleStorageError(err error, entityName string) error {
+	ErrPlanMustBeBindable                    = errors.New("plan must be bindable")
+	ErrAsyncNotSupportedForSharing           = errors.New("can't use async for sharing instances")
+	ErrReferencedInstanceNotShared           = errors.New("referenced-instance should be shared first")
+	ErrChangingPlanOfReferenceInstance       = errors.New("changing plan of reference instance")
+	ErrChangingPlanOfSharedInstance          = errors.New("changing plan of shared instance")
+	ErrChangingParametersOfReferenceInstance = errors.New("changing parameters of reference instance")
+	ErrMissingOrInvalidReferenceParameter    = errors.New("missing or invalid referenced_instance_id parameter")
+	ErrParsingNewCatalogWithReference        = errors.New("failed generating reference-plan")
+	ErrUnknownOSBMethod                      = errors.New("osb method is unknown")
+	ErrSharedPlanHasReferences               = errors.New("shared plan has references")
+	ErrInstanceIsAlreadyAtDesiredSharedState = errors.New("instance already at the desired shared state")
+	ErrInvalidShareRequest                   = errors.New("invalid share request")
+)
+
+func HandleInstanceSharingError(err error, entityName string) error {
+	if err == nil {
+		return nil
+	}
+
+	if _, ok := err.(*HTTPError); ok {
+		return err
+	}
+
+	switch err {
+	case ErrCatalogUsesReservedPlanName:
+		return &HTTPError{
+			ErrorType:   "BadRequest",
+			Description: fmt.Sprintf("The plan name you used for registration \"%s\" is reserved for the Service Manager; you must choose a different name.", entityName),
+			StatusCode:  http.StatusBadRequest,
+		}
+	case ErrPlanMustBeBindable:
+		return &HTTPError{
+			ErrorType:   "BadRequest",
+			Description: fmt.Sprintf("The plan %s must be defined as 'bindable' so that it can support instance sharing.", entityName),
+			StatusCode:  http.StatusBadRequest,
+		}
+	case ErrReferencedInstanceNotShared:
+		return &HTTPError{
+			ErrorType:   "BadRequest",
+			Description: fmt.Sprintf("Failed to create the reference. The instance %s, for which you want to create the reference, must be shared first.", entityName),
+			StatusCode:  http.StatusBadRequest,
+		}
+	case ErrChangingPlanOfReferenceInstance:
+		return &HTTPError{
+			ErrorType:   "BadRequest",
+			Description: fmt.Sprintf("Failed to update the instance %s. This is a reference instance, therefore its plan can't be changed.", entityName),
+			StatusCode:  http.StatusBadRequest,
+		}
+	case ErrChangingPlanOfSharedInstance:
+		return &HTTPError{
+			ErrorType:   "BadRequest",
+			Description: fmt.Sprintf("Couldn't update the instance's plan. The instance %s is shared, therefore you must unshare it first.", entityName),
+			StatusCode:  http.StatusBadRequest,
+		}
+	case ErrChangingParametersOfReferenceInstance:
+		return &HTTPError{
+			ErrorType:   "BadRequest",
+			Description: fmt.Sprintf("Failed to update the instance %s. This is a reference instance, therefore its parameters can't be changed.", entityName),
```

Comment:
There is only one parameter, reference instance id, which is the pointer to the root instance. we can be more specific.
e.g. "It is not allowed to change the referenced instance"

Reply:
In that case, we block any change of parameters.
I agree it is a generic error message, but when we validate it, it can be any parameter.

Reply:
For some of these, it's a matter of taste.
I am not a fan of "it is not allowed", I like to be a bit less negative.

Reply:
Ok, then can we go with:
"Failed to update the instance %s. It isn't allowed to change a reference instance."

Reply:
changed.